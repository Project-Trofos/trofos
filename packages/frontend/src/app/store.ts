import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
