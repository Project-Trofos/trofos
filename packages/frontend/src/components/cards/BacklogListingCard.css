li.backlog-card-container.ant-list-item {
  background: #fff;
  padding: 0;
  border-block-end: none;
}

.backlog-card-container > div > div {
  margin-right: 8px;
}

.backlog-card-id {
  word-break: break-all;
  width: 60px;
  margin-top: 4px;
}
