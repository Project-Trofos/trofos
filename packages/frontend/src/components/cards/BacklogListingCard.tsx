import React from 'react';
import { isEqual } from 'lodash';
import type { Backlog } from '../../api/types';
import { UserData } from '../../api/types';
import BacklogCardStatus from '../dropdowns/BacklogCardStatus';
import BacklogCardType from '../dropdowns/BacklogCardType';
import BacklogCardPriority from '../dropdowns/BacklogCardPriority';
import BacklogCardAssignee from '../dropdowns/BacklogCardAssignee';
import BacklogCardPoints from '../fields/BacklogCardPoints';
import BacklogCardSummary from '../fields/BacklogCardSummary';
import './BacklogListingCard.css';
import BacklogCardEpic from '../dropdowns/BacklogCardEpic';

function BacklogListingCard(props: BacklogListingCardProps): JSX.Element {
  const { backlog, projectKey, users } = props;

  return (
    <>
      <div className="backlog-card-id">
        {projectKey ? `${projectKey}-` : ''}
        {backlog.backlog_id}
      </div>
      <BacklogCardSummary backlogId={backlog.backlog_id} currentSummary={backlog.summary} />
      <BacklogCardStatus backlogId={backlog.backlog_id} currentStatus={backlog.status} />
      <BacklogCardType backlogId={backlog.backlog_id} currentType={backlog.type} />
      <BacklogCardPriority backlogId={backlog.backlog_id} currentPriority={backlog.priority} />
      <BacklogCardAssignee backlogId={backlog.backlog_id} currentAssignee={backlog.assignee_id} projectUsers={users} />
      <BacklogCardPoints backlogId={backlog.backlog_id} currentPoints={backlog.points} />
      <BacklogCardEpic backlogId={backlog.backlog_id} currentEpic={backlog.epic_id ?? undefined} />
    </>
  );
}

type BacklogListingCardProps = {
  backlog: Backlog;
  projectKey: string | null | undefined;
  users?: UserData[];
};

export default React.memo(BacklogListingCard, (prevProps, currProps) => isEqual(prevProps, currProps));
