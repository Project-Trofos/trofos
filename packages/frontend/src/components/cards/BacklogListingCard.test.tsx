import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import BacklogListingCard from './BacklogListingCard';
import type { Backlog } from '../../api/backlog';
import store from '../../app/store';

describe('BacklogCard test', () => {
  const mockBacklog: Backlog = {
    backlog_id: 111,
    summary: 'A Test Summary',
    type: 'story',
    priority: 'very_high',
    sprint_id: 123,
    reporter_id: 1,
    assignee_id: 2,
    points: 3,
    description: 'A test description here',
    project_id: 123,
    status: 'todo',
  };
  const { baseElement } = render(
    <Provider store={store}>
      <BacklogListingCard backlog={mockBacklog} projectKey="MOCK" />
    </Provider>,
  );

  it('renders backlog card with correct details', () => {
    expect(screen.getByText('MOCK-111')).toBeInTheDocument();
    expect(screen.getByDisplayValue('A Test Summary')).toBeInTheDocument();
    expect(screen.getByText('Story')).toBeInTheDocument();
    expect(screen.getByText('Very High')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument();
    expect(screen.getByDisplayValue(3)).toBeInTheDocument();

    // Compare with snapshot to ensure structure remains the same
    expect(baseElement).toMatchSnapshot();
  });
});
