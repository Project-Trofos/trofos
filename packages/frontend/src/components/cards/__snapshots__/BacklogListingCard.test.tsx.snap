// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BacklogCard test > renders backlog card with correct details 1`] = `
<body>
  <div>
    <div
      class="ant-space ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
      style="width: 100%; padding: 16px; background-color: rgb(255, 255, 255); border-radius: 8px; border: 1px solid #d9d9d9; overflow-x: auto;"
    >
      <div
        class="ant-space-item"
      >
        <div
          class="backlog-card-id"
        >
          MOCK-
          111
        </div>
      </div>
      <div
        class="ant-space-item"
      >
        <textarea
          class="ant-input ant-input-outlined backlog-card-summary"
          style="resize: none; height: 2px; min-height: 2px;"
        >
          A Test Summary
        </textarea>
      </div>
      <div
        class="ant-space-item"
      >
        <div
          class="ant-select ant-select-outlined backlog-card-status ant-select-single ant-select-show-arrow"
        >
          <div
            class="ant-select-selector"
          >
            <span
              class="ant-select-selection-wrap"
            >
              <span
                class="ant-select-selection-search"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="ant-select-selection-item"
                title="todo"
              >
                todo
              </span>
            </span>
          </div>
          <span
            aria-hidden="true"
            class="ant-select-arrow"
            style="user-select: none;"
            unselectable="on"
          />
        </div>
      </div>
      <div
        class="ant-space-item"
      >
        <div
          class="ant-select ant-select-outlined backlog-card-type ant-select-single ant-select-show-arrow"
        >
          <div
            class="ant-select-selector"
          >
            <span
              class="ant-select-selection-wrap"
            >
              <span
                class="ant-select-selection-search"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="ant-select-selection-item"
              >
                <div
                  class="ant-space ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
                >
                  <div
                    class="ant-space-item"
                  >
                    <span
                      aria-label="wallet"
                      class="anticon anticon-wallet"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="wallet"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 464H528V448h312v128zm0 264H184V184h656v200H496c-17.7 0-32 14.3-32 32v192c0 17.7 14.3 32 32 32h344v200zM580 512a40 40 0 1080 0 40 40 0 10-80 0z"
                        />
                      </svg>
                    </span>
                  </div>
                  <div
                    class="ant-space-item"
                  >
                    Story
                  </div>
                </div>
              </span>
            </span>
          </div>
          <span
            aria-hidden="true"
            class="ant-select-arrow"
            style="user-select: none;"
            unselectable="on"
          />
        </div>
      </div>
      <div
        class="ant-space-item"
      >
        <div
          class="ant-select ant-select-outlined backlog-card-priority very_high-priority ant-select-single ant-select-allow-clear ant-select-show-arrow"
        >
          <div
            class="ant-select-selector"
          >
            <span
              class="ant-select-selection-wrap"
            >
              <span
                class="ant-select-selection-search"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="ant-select-selection-item"
                title="Very High"
              >
                Very High
              </span>
            </span>
          </div>
          <span
            aria-hidden="true"
            class="ant-select-arrow"
            style="user-select: none;"
            unselectable="on"
          />
          <span
            aria-hidden="true"
            class="ant-select-clear"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              aria-label="close-circle"
              class="anticon anticon-close-circle"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="close-circle"
                fill="currentColor"
                fill-rule="evenodd"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm127.98 274.82h-.04l-.08.06L512 466.75 384.14 338.88c-.04-.05-.06-.06-.08-.06a.12.12 0 00-.07 0c-.03 0-.05.01-.09.05l-45.02 45.02a.2.2 0 00-.05.09.12.12 0 000 .07v.02a.27.27 0 00.06.06L466.75 512 338.88 639.86c-.05.04-.06.06-.06.08a.12.12 0 000 .07c0 .03.01.05.05.09l45.02 45.02a.2.2 0 00.09.05.12.12 0 00.07 0c.02 0 .04-.01.08-.05L512 557.25l127.86 127.87c.04.04.06.05.08.05a.12.12 0 00.07 0c.03 0 .05-.01.09-.05l45.02-45.02a.2.2 0 00.05-.09.12.12 0 000-.07v-.02a.27.27 0 00-.05-.06L557.25 512l127.87-127.86c.04-.04.05-.06.05-.08a.12.12 0 000-.07c0-.03-.01-.05-.05-.09l-45.02-45.02a.2.2 0 00-.09-.05.12.12 0 00-.07 0z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
      <div
        class="ant-space-item"
      >
        <div
          class="ant-select ant-select-outlined backlog-card-assignee ant-select-single ant-select-allow-clear ant-select-show-arrow"
        >
          <div
            class="ant-select-selector"
          >
            <span
              class="ant-select-selection-wrap"
            >
              <span
                class="ant-select-selection-search"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="ant-select-selection-item"
                title="2"
              >
                2
              </span>
            </span>
          </div>
          <span
            aria-hidden="true"
            class="ant-select-arrow"
            style="user-select: none;"
            unselectable="on"
          />
          <span
            aria-hidden="true"
            class="ant-select-clear"
            style="user-select: none;"
            unselectable="on"
          >
            <span
              aria-label="close-circle"
              class="anticon anticon-close-circle"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="close-circle"
                fill="currentColor"
                fill-rule="evenodd"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm127.98 274.82h-.04l-.08.06L512 466.75 384.14 338.88c-.04-.05-.06-.06-.08-.06a.12.12 0 00-.07 0c-.03 0-.05.01-.09.05l-45.02 45.02a.2.2 0 00-.05.09.12.12 0 000 .07v.02a.27.27 0 00.06.06L466.75 512 338.88 639.86c-.05.04-.06.06-.06.08a.12.12 0 000 .07c0 .03.01.05.05.09l45.02 45.02a.2.2 0 00.09.05.12.12 0 00.07 0c.02 0 .04-.01.08-.05L512 557.25l127.86 127.87c.04.04.06.05.08.05a.12.12 0 00.07 0c.03 0 .05-.01.09-.05l45.02-45.02a.2.2 0 00.05-.09.12.12 0 000-.07v-.02a.27.27 0 00-.05-.06L557.25 512l127.87-127.86c.04-.04.05-.06.05-.08a.12.12 0 000-.07c0-.03-.01-.05-.05-.09l-45.02-45.02a.2.2 0 00-.09-.05.12.12 0 00-.07 0z"
                />
              </svg>
            </span>
          </span>
        </div>
      </div>
      <div
        class="ant-space-item"
      >
        <div
          class="ant-input-number ant-input-number-outlined backlog-card-points points-active"
        >
          <div
            class="ant-input-number-input-wrap"
          >
            <input
              aria-valuemin="1"
              aria-valuenow="3"
              autocomplete="off"
              class="ant-input-number-input"
              placeholder="points"
              role="spinbutton"
              step="1"
              value="3"
            />
          </div>
        </div>
      </div>
      <div
        class="ant-space-item"
      >
        <div
          class="ant-select ant-select-outlined backlog-epic ant-select-single ant-select-allow-clear ant-select-show-arrow"
        >
          <div
            class="ant-select-selector"
          >
            <span
              class="ant-select-selection-wrap"
            >
              <span
                class="ant-select-selection-search"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="rc_select_TEST_OR_SSR_list"
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-owns="rc_select_TEST_OR_SSR_list"
                  autocomplete="off"
                  class="ant-select-selection-search-input"
                  id="rc_select_TEST_OR_SSR"
                  readonly=""
                  role="combobox"
                  style="opacity: 0;"
                  type="search"
                  unselectable="on"
                  value=""
                />
              </span>
              <span
                class="ant-select-selection-placeholder"
              >
                epic
              </span>
            </span>
          </div>
          <span
            aria-hidden="true"
            class="ant-select-arrow"
            style="user-select: none;"
            unselectable="on"
          />
        </div>
      </div>
    </div>
  </div>
  <textarea
    aria-hidden="true"
    name="hiddenTextarea"
    style="letter-spacing:normal;line-height:1.5714285714285714;padding-top:4px;padding-bottom:4px;font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,'Helvetica Neue',Arial,'Noto Sans',sans-serif,'Apple Color Emoji','Segoe UI Emoji','Segoe UI Symbol','Noto Color Emoji';font-weight:;font-size:14px;font-variant:;text-rendering:auto;text-transform:none;width:100%;text-indent:0;padding-left:11px;padding-right:11px;border-width:1px;box-sizing:border-box;word-break:break-all;white-space:pre-wrap;
  min-height:0 !important;
  max-height:none !important;
  height:0 !important;
  visibility:hidden !important;
  overflow:hidden !important;
  position:absolute !important;
  z-index:-1000 !important;
  top:0 !important;
  right:0 !important;
  pointer-events: none !important;
"
    tab-index="-1"
  />
</body>
`;
